### PR TITLE
from internal working branch

### PR DESCRIPTION
**Pull request by bkrantz**
Changes include ...

Implemented error handling override,
Implemented small flow controlling changes,
Removed blanket excepts
Implemented iterables where possible
Updated setup.py file

### CODE REVIEW

File: compysition/actor.py [hunk 1]
```diff
@@ -186,16 +186,15 @@ def send_event(self, event, queues=__NOT_DEFINED, check_output=True):
         """
 
         if queues is self.__NOT_DEFINED:
```
Comment (jameshulett) [inline compysition/actor.py]: It is generally more pythonic to use None as opposed to a sentinel value such as __NOT_DEFINED. You can then alter your `if` statement above as commented.

Comment (jameshulett) [inline compysition/actor.py]: if queues is self.__NOT_DEFINED can become if not queues

File: compysition/actor.py [hunk 2]
```diff
@@ -206,9 +205,12 @@ def _loop_send(self, event, queues, check_output=True):
         if check_output and not isinstance(event, self.output):
             raise InvalidActorOutput("Event was of type '{_type}', expected '{output}'".format(_type=type(event), output=self.output))
 
-        if len(queues) > 0:
-            self._send(queues[0], deepcopy(event))
-            map(lambda _queue: self._send(_queue, deepcopy(event)), queues[1:])
+        if isinstance(queues, _InternalQueuePool):
```
Comment (jameshulett) [inline compysition/actor.py]: I would look into a try/except block instead of isinstance. Python style is usually forgiveness as opposed to permission.

File: compysition/actors/basicauth.py [hunk 3]
```diff
@@ -37,7 +37,7 @@ def consume(self, event, *args, **kwargs):
             try:
                 authorization = event.environment['HTTP_AUTHORIZATION']
                 user, password = base64.decodestring(authorization.split(' ')[1]).split(':')
-            except:
+            except Exception:
```
Comment (jameshulett) [inline compysition/actors/basicauth.py]: I believe `except` and `except Exception` are the same net effect. I would recommend catching `KeyError` for the case of "No auth headers..."

File: compysition/actors/eventattributemodifier.py [hunk 4]
```diff
@@ -177,4 +177,141 @@ def get_modify_value(self, event):
...
+                        item = keys.pop(0)
+                        next_step = current_element.get(item, None)
+                        if not next_step:
+                            next_step = etree.Element(item)
+                            current_element.append(next_step)
+                        current_element = next_step
+                    else:
+                        current_element.text = str(value)
+                        break
+            except Exception as err:
```
Comment (jameshulett) [inline compysition/actors/eventattributemodifier.py]: You generally want to consume only the exception to which your log message applies and let the rest bubble up or get caught.

File: compysition/actors/eventattributemodifier.py [hunk 5]
```diff
@@ -177,4 +177,141 @@ def get_modify_value(self, event):
...
+                                                                If the variable is missing then a static message is used if provided.
+        - error_class (CompysitionError class):             (See ErrorEventAttributeModifier)
+        - default_error_class (CompysitionError class):     (See ErrorEventAttributeModifier)
+    '''
+    def __init__(self, name, override=False, code=None, code_attr=None, message=None, message_attr=None, error_class=None, default_error_class=MalformedEventData, *args, **kwargs):
+        self.message_attr = message_attr
+        self.code_attr = code_attr
+        super(XMLErrorEventAttributeModifier, self).__init__(name=name, override=override, code=code, message=message, error_class=error_class, default_error_class=default_error_class, *args, **kwargs)
+
+    def _get_attr(self, event, attr):
```
Comment (jameshulett) [inline compysition/actors/eventattributemodifier.py]: Would this maybe be more efficient converting to a dict using `json` from the standard lib?

File: compysition/actors/httpserver.py [hunk 6]
```diff
@@ -267,8 +268,8 @@ def consume(self, event, *args, **kwargs):
             status, status_message = event.status
             local_response.status = "{code} {message}".format(code=status, message=status_message)
 
-            for header in event.headers.keys():
-                local_response.set_header(header, event.headers[header])
+            for header, value in event.headers.iteritems():
```
Comment (jameshulett) [inline compysition/actors/httpserver.py]: You... rule

File: compysition/actors/httpserver.py [hunk 7]
```diff
@@ -292,7 +293,7 @@ def _format_bottle_env(self, environ):
         for key in environ["bottle.request"].query.iterkeys():
             query_string_data[key] = environ["bottle.request"].query.get(key)
 
-        environ = {key: environ[key] for key in environ.keys() if isinstance(environ[key], (str, tuple, bool, dict))}
+        environ = {key: environ[key] for key in environ.iterkeys() if isinstance(environ[key], (str, tuple, bool, dict))}
```
Comment (jameshulett) [inline compysition/actors/httpserver.py]: Again with the ruling...

File: compysition/actors/util/udplib.py [hunk 8]
```diff
@@ -178,15 +178,15 @@ def run(self):
     
     def reap_peers(self):
         now = time.time()
-        for peer in list(self.peers.values()):
-            if peer.expires_at < now:
-                self.peers.pop(peer.uuid)
-                self.log("Reaping expired peer ({0}) from '{1}' peer pool. (Total remaining: {2})".format(peer.uuid, self.service, len(self.peers) + 1))
+        remove = [peer for peer in self.peers.itervalues() if peer.expires_at < now]
```
Comment (jameshulett) [inline compysition/actors/util/udplib.py]: Could this be a generator instead of a list?

File: compysition/director.py [hunk 9]
```diff
@@ -145,9 +145,8 @@ def finalize_blockdiag(self):
             if not os.path.exists(img_dir):
                 os.makedirs(img_dir)
             from blockdiag.command import BlockdiagApp
-            f = open("{0}{1}{2}.diag".format(self.blockdiag_dir, os.sep, self.name),'w')
-            f.write(self.blockdiag_out)
-            f.close()
+            with open("{0}{1}{2}.diag".format(self.blockdiag_dir, os.sep, self.name),'w') as f
```
Comment (jameshulett) [inline compysition/director.py]: In languages that express type, names like f are legit. if I have for example `FileObjects f<*File>` I know f stands  for file. In Python, we don't have types, so names are a bit more important for clarity. Fine here, just get's more important as code does complex.

File: compysition/event.py [hunk 10]
```diff
@@ -204,15 +220,16 @@ def _set_error(self, exception):
...
-                         messages)
-            return errors
-
+            if self.error.override:
+                return self.error.override
+            else:
+                messages = self.error.message
+                if not isinstance(messages, list):
+                    messages = [messages]
+                errors = map(lambda _error:
```
Comment (jameshulett) [inline compysition/event.py]: Could this be a dictionary comprehension or ChainMap?

File: compysition/queue.py [hunk 11]
```diff
@@ -73,14 +73,19 @@ def __init__(self, size=0):
...
+            self.outbound,
+            self.error,
+            self.logs
+        ]
 
     def join(self):
         """**Blocks until all queues in the pool are empty.**"""
-        for queue in self.list_all_queues():
-            queue.wait_until_empty()
+        for lst in self._list_of_queue_lists():
```
Comment (jameshulett) [inline compysition/queue.py]: Was this change for efficiency?

File: setup.py [hunk 12]
```diff
@@ -48,26 +52,24 @@
...
 setup(
     name=PROJECT,
     version=VERSION,
     description='Build event pipeline servers with minimal effort.',
     long_description=long_description,
-    author='Adam Fiebig',
-    author_email='fiebig.adam@gmail.com',
-    url='https://github.com/fiebiga/compysition',
-    download_url='https://github.com/fiebiga/compysition/tarball/master',
+    author='Integrations Team',
```
Comment (jameshulett) [inline setup.py]: I would let Don tell him

File: setup.py [hunk 13]
```diff
@@ -48,26 +52,24 @@
...
+    author_email='integrations@cuanswers',
```
Comment (drpoggi) [inline setup.py]: .com?

File: compysition/actor.py (hunk 2 above)
Comment (bkrantz) [inline compysition/actor.py]: I'm not opposed, however my thought here was that I didn't want just any dict to work, only _InternalQueuePool.

Comment (jameshulett) [inline compysition/actor.py]: Why not? What if you extended `_InternalQueuePool` and wanted to inject another dependency so to speak. Python's MRO is not linear. `isinstance` in in the language for a reason. It's use case is just rare enough for me to wonder if try catch would be more flexible and idiomatic. I admit I do not have all the background.

File: compysition/actors/basicauth.py (hunk 3 above)
Comment (bkrantz) [inline compysition/actors/basicauth.py]: For all try/except blocks I agree I would like to be very specific as far was what Exception to be caught.  I haven't done this yet as I wasn't willing to invest the time to determine the desired Exceptions.  However I do not believe "except" and "except Exception" are the same.  I believe the later is still an improvement, as it won't catch unintended Exceptions like KeyboardInterupt.

File: compysition/actors/eventattributemodifier.py (hunk 5 above)
Comment (bkrantz) [inline compysition/actors/eventattributemodifier.py]: vs converting the event as on line 301 (below)?

File: compysition/actors/util/udplib.py (hunk 8 above)
Comment (bkrantz) [inline compysition/actors/util/udplib.py]: The issue here is the pop.  There may be a way to use a generator, but an iterator could fail.

File: compysition/queue.py (hunk 11 above)
Comment (bkrantz) [inline compysition/queue.py]: I don't like this change as I am removing a function all together and it feels "hacky" (and haven't fully tested).  Though the theory was to iterate each list rather than iterating a **new** merged list.

File: setup.py (hunk 12 above)
Comment (bkrantz) [inline setup.py]: Yeah I'm not a fan of confrontation ... to much stress

File: compysition/actor.py (hunk 2 above)
Comment (bkrantz) [inline compysition/actor.py]: Honestly, I was afraid of dicts with non queue object values containing a put attribute.  Thinking that could potentially lead to undesired (non error) side effects.  Though the more I think about it, that is an edge case, and one would have to go out of there way to make that happen.  So .. I am now much more inclined to swap isinstance with try as, I agree, pythonically it is better to ask forgiveness.

File: compysition/actors/util/udplib.py [hunk 14]
```diff
@@ -178,15 +178,12 @@ def run(self):
     
     def reap_peers(self):
         now = time.time()
-        for peer in list(self.peers.values()):
-            if peer.expires_at < now:
-                self.peers.pop(peer.uuid)
-                self.log("Reaping expired peer ({0}) from '{1}' peer pool. (Total remaining: {2})".format(peer.uuid, self.service, len(self.peers) + 1))
```
Comment (drpoggi) [inline compysition/actors/util/udplib.py]: We should probably keep this log message

File: compysition/actors/zeromq.py [hunk 15]
```diff
@@ -96,7 +96,7 @@ def __init__(self, name, port=DEFAULT_PORT, transmission_protocol=TCP, socket_fi
                                     self.IPC: "ipc://{0}".format(socket_file),
                                     self.INPROC: "inproc://{0}".format(socket_file)}
 
-        if transmission_protocol in self.format_connection.keys():
+        if self.format_connection.get(transmission_protocol, None):
```
Comment (drpoggi) [inline compysition/actors/zeromq.py]: Why was this changed? I think ideally this would be `if transmission_protocol in self.format_connection`, we can drop the `.keys()`. This also changes what we're doing (not in practice, it still works fine obviously), as instead of checking if the key is in the dict, we're evaluating the value of that key.

File: compysition/event.py [hunk 16]
```diff
@@ -159,14 +157,32 @@ def event_id(self, event_id):
...
     def lookup(self, path):
         """
-        TODO: Account for list objects in the lookup path and generate multiple results if found
+        Implements the retrieval of a single list index through an integer path entry
         """
         if isinstance(path, str):
             path = [path]
 
-        value = reduce(lambda obj, key: obj.get(key, NullLookupValue()) if isinstance(obj, dict) else getattr(obj, key, NullLookupValue()), [self] + path)
+        value = reduce(lambda obj, key: self._obj_get(obj, key, self._getattr(obj, key, self._list_get(obj, key, NullLookupValue()))), [self] + path)
```
Comment (drpoggi) [inline compysition/event.py]: Can you fill me in on why this was changed?

Comment (bkrantz) [reply compysition/event.py]: To implement the retrieval of a list item and to limit the use of isinstance.  Also, I guess I didn't catch it before but, eventually I would like to replace the NullLookupValue with None.

File: compysition/actors/zeromq.py (hunk 15 above)
Comment (bkrantz) [reply compysition/actors/zeromq.py]: My initial goal was simply to remove the `.keys()`.  When I made this change I was thinking the transmission_protocol value should never be None but after further review of the context it looks like it never will be.  So, I agree `if transmission_protocol in self.format_connection` is probably more efficient.

File: compysition/actors/util/udplib.py (hunk 14 above)
Comment (bkrantz) [reply compysition/actors/util/udplib.py]: I understand wanting to keep it.  I was also hesitant to remove it.  However I can't see a way where we would not be sacrificing efficiency to have it in there.  What are your thoughts?

Comment (drpoggi) [reply compysition/actors/util/udplib.py]: Are there significant differences in efficiency? Worth losing a log message over?

Comment (bkrantz) [reply compysition/actors/util/udplib.py]: Really i don't think it would be the end of the world if we left it the same, though I honestly don't know how important that log message is?

The way I see it, least to most efficient, we can either leave it as it was, incorporate itervalue and a temp variable (with log), or use a dictionary comprehension (without log).
